### ISSUE TITLE
quick explanation of pytorch lstm

### ISSUE THREAD

**Issue by sinhasam** [yunjey/pytorch-tutorial#40]
Hi, 

I have been wondering this for a while; in https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/02-intermediate/recurrent_neural_network/main.py, why are you setting new hidden and cell states for each forward propagation? Why are h0 and c0 not being passed in as arguments to the forward function and returned after the forward prop. Example: 

```
h0 = nn.Parameter(torch.randn(num_layers*2, 100, hidden_size)) # 2 for bidirection 
c0 = nn.Parameter(torch.randn(num_layers*2, 100, hidden_size))
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        images = Variable(images.view(-1, sequence_length, input_size))
        labels = Variable(labels)
        
        optimizer.zero_grad()
        outputs, h0, c0 = rnn.forward(images, h0, c0)

```

 Dont the partial derivatives of the hidden and cell states also effect the accuracy of the model?

**Comment by yunjey** [comment]
```python
def forward(self, x):
    # Set initial states 
    h0 = Variable(torch.zeros(self.num_layers, x.size(0), self.hidden_size)) 
    c0 = Variable(torch.zeros(self.num_layers, x.size(0), self.hidden_size))
        
    # Forward propagate RNN
    out, (h_last, c_last) = self.lstm(x, (h0, c0))    # same as self.lstm(x). Default is zero initialization.
        
    # Decode hidden state of last time step
    out = self.fc(out[:, -1, :])             # same as out = h_last
    return out
```


@sinhasam  You should note that `images` is a 3D tensor of shape **(batch_size, seq_length, input_size)**.  In `out, (h_last, c_last) = lstm(x, (h0, c0))`, `out` is a tensor of shape **(batch_size, seq_length, hidden_size)** and `h_last `and `c_last `are tensors of shape **(batch_size, hidden_size)** which indicate the last hidden and cell state of the lstm (Full time step forward propagation, not one-step). You can alse see [here]((http://pytorch.org/docs/nn.html#lstm)) to understand the mechanism of nn.LSTM.


In addition, in many-to-one case, we need only last hidden state of the lstm. For this, we can use `out[:, -1, :]` or `h_last`. In our case (MNIST), each sequence in `x` has a fixed length of 28 so `output[:, -1, :]` and `h_last` are exactly same.   For variable length many-to-one, you can use [pack_padded_sequence](http://pytorch.org/docs/nn.html?highlight=pack_padded_sequence#torch.nn.utils.rnn.pack_padded_sequence) and `h_last`. Please, see [here](https://discuss.pytorch.org/t/lstm-hidden-cell-outputs-and-packed-sequence-for-variable-length-sequence-inputs/1183) for the details.
